src/credentials/MessageSecurity.cpp

```diff
@@ -0,0 +1,237 @@
+/*
+ *
+ *    Copyright (c) 2020-2022 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *  @file Implements message encryption and decryption
+ */
+
+#include "MessageSecurity.h"
+#include <credentials/GroupDataProvider.h>
+#include <lib/support/SafeInt.h>
+
+using namespace chip::Credentials;
+
+namespace chip {
+
+namespace MessageSecurity {
+
+constexpr size_t kAESCCMIVLen = 13;
+constexpr size_t kMaxAADLen   = 128;
+
+CHIP_ERROR GetIV(const chip::PacketHeader & header, uint8_t * iv, size_t len)
```

Comment:
This is duplicating `CryptoContext::GetIV`, no?  Why are we not sharing the crypto code?

Reply:
All the encryption-decryption functions is copy-pasted. Unfortunately, the cryptography code is located under 'transport', and due to multiple dependencies, is not reusable by the GroupDataProvider.
I can modify the transport code to reuse this file instead.

Reply:
`src/credentials` can't use things in `src/transport`?  That seems odd.  @mspang @tcarmelveilleux @pan-apple 

In particular, this code is using `PacketHeader`, which is under `src/transport`.  What specific dependency problems arise when this code tries to use the existing crypto bits?